### PR TITLE
ATT-27: Modify the upload method in rest API

### PR DESCRIPTION
**Pull request by ridmal**
[ATT-24 ](https://issues.openmrs.org/browse/ATT-27) 

REST end point for file upload should allow to specify encounter

### CODE REVIEW

**Comment by dkayiwa** [comment]
@mks-d what do you think of this?

File: omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentController1_10Test.java [hunk 1]
```diff
@@ -247,6 +248,34 @@ public void postAttachment_shouldUploadFile() throws Exception {
...
+			
+			Obs obs = Context.getObsService().getObsByUuid((String) response.get("uuid"));
+			Obs complexObs = Context.getObsService().getComplexObs(obs.getObsId(), null);
+			ComplexData complexData = complexObs.getComplexData();
+			
+			// Verify
+			Assert.assertEquals(obs.getComment(), fileCaption);
+			Assert.assertEquals(complexData.getTitle(), fileName);
+			Assert.assertArrayEquals(randomData, (byte[]) complexData.getData());
+			Assert.assertNotNull(obs.getEncounter());
```
Comment (mks-d) [inline omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentController1_10Test.java]: You should verify that it is the expected encounter, so the one whose UUID is `encounter.getUuid()`.

**Review by mks-d — CHANGES_REQUESTED** [review]
@ridmal this is looking good, thank you.

But because there are several possible cases now, I would like to split this in several ad-hoc test methods:

1. `postAttachment_shouldUploadFileToVisit`
1. `postAttachment_shouldUploadFileAsEncounterless`
1. `postAttachment_shouldUploadFileToEncounter`

Would you mind doing this as your next commit on this branch?

File: omod-1.10/src/main/java/org/openmrs/module/attachments/rest/AttachmentResource1_10.java [hunk 2]
```diff
@@ -103,11 +104,21 @@ public Object upload(MultipartFile file, RequestContext context) throws Response
 		if (StringUtils.isEmpty(instructions))
 			instructions = ValueComplex.INSTRUCTIONS_DEFAULT;
 		
-		Encounter encounter = null;
-		if (visit != null) {
+		// Verify Parameters
+		if (encounter != null && visit != null) {
+			if (encounter.getVisit() != visit) {
+				throw new IllegalRequestException("Visit and Encounter cannot be mismatched");
```
Comment (mks-d) [inline omod-1.10/src/main/java/org/openmrs/module/attachments/rest/AttachmentResource1_10.java]: Make the error message:
>`"The specified encounter does not belong to the provided visit, upload aborted."`

File: omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentController1_10Test.java [hunk 3]
```diff
@@ -247,6 +248,56 @@ public void postAttachment_shouldUploadFile() throws Exception {
...
```
Comment (mks-d) [inline omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentController1_10Test.java]: Please address my above comment.

File: omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentController1_10Test.java [hunk 4]
```diff
@@ -247,6 +248,56 @@ public void postAttachment_shouldUploadFile() throws Exception {
...
+		}
+	}
+	
+	@Test(expected = IllegalRequestException.class)
+	public void postAttachment_shouldNotUploadFileWithDifferentEncounterAndVisit() throws Exception {
```
Comment (mks-d) [inline omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentController1_10Test.java]: → `postAttachment_shouldThrowWhenVisitAndEncounterDoNotMatch`

**Review by mks-d — COMMENTED** [review]
A few comments, but great!

Please resume based on my comments from yesterday, in particular to provide more test cases and split them in to separate methods.

**Comment by ridmal** [comment]
Thanks for your feedbacks @mks-d  and @dkayiwa . I update the PR .

**Pull request merged**